chore(CI/CD-GitLab): jobs templates and examples (#1944)
mendersoftware/mender-docs: 06.Artifact-creation/10.CI-CD/02.GitLab-CICD/docs.md

    label: tutorial
---

## Job templates
<!--AUTOVERSION: "tree/%/templates"/ignore-->

> In addition to the [mender-ci-tools Docker image](../docs.md#mender-ci-workflows-docker-image) Mender provides [job templates](https://github.com/mendersoftware/mender-ci-workflows/tree/master/templates/gitlab) for uploading Mender Artifacts and creating deployments to a group of devices.

Reviewer: ```suggestion
In addition to the [mender-ci-tools Docker image](../docs.md#mender-ci-workflows-docker-image), Mender provides [job templates](https://github.com/mendersoftware/mender-ci-workflows/tree/master/templates/gitlab) for uploading Mender Artifacts and creating deployments to a group of devices.
```

<!--AUTOVERSION: "Mender server version % or"/ignore-->
!!!!! The jobs use [Personal Access Tokens](../../../08.Server-integration/01.Using-the-apis/docs.md#personal-access-tokens) feature which is only available in Mender server version 3.4 or newer.

In this chapter we introduce the different jobs and provide [examples](#pipelines-examples) on how to use them.

### Upload Mender Artifact job

> <!--AUTOVERSION: "tree/%/templates"/ignore-->

Reviewer: These cannot be `ignore`, they should follow the target repo:
```suggestion
<!--AUTOVERSION: "tree/%/templates"/mender-ci-workflows-->
```

I privately commented about this for the [previous PR](https://github.com/mendersoftware/mender-docs/pull/1937) but I missed it in your rework. Could you please fix also for the GitHub Actions doc page?

[.mender:upload:artifact](https://github.com/mendersoftware/mender-ci-workflows/tree/master/templates/gitlab/mender-artifact-upload.gitlab-ci.yml) job template uploads a Mender Artifact to a Mender server.

### Create deployment job
<!--AUTOVERSION: "tree/%/templates"/ignore-->
[.mender:create:deployment](https://github.com/mendersoftware/mender-ci-workflows/tree/master/templates/gitlab/mender-deployment-create.gitlab-ci.yml) job template creates a deployment on a Mender server.

### Setup

> The actions requires the following secret is set in a repository settings:

Reviewer: Templates, no actions
```suggestion
The job templates require the following secret is set in a repository settings:
```

And I suggest to be more explicit, as GitLab UI does not have "secrets" only variables
```suggestion
The actions requires the following CI variable to be set in the repository settings:

!!!!!! Make sure to protect the variable accordingly
```

Reviewer: I suggest we add a note or warning like above:

`!!! Make sure to protect the variable accordingly`

- `MENDER_SERVER_ACCESS_TOKEN`: Mender [Personal Access Token](../../../08.Server-integration/01.Using-the-apis/docs.md#personal-access-tokens)

## Pipelines examples

### Build and deploy Mender Artifact
<!--AUTOVERSION: "tree/%/examples"/ignore-->
```bash
@build-and-deploy-mender-artifact.gitlab-ci.yml@ # https://github.com/mendersoftware/mender-ci-workflows/tree/master/examples/gitlab/build-and-deploy-mender-artifact.gitlab-ci.yml
```

### Build and deploy multiple Mender Artifacts
<!--AUTOVERSION: "tree/%/examples"/ignore-->
```bash
@build-two-artifacts.gitlab-ci.yml@ # https://github.com/mendersoftware/mender-ci-workflows/tree/master/examples/gitlab/build-two-artifacts.gitlab-ci.yml

> ```

Reviewer: Didn't we agree on leaving this one out?

Author: I don't remember it. Do you think it's reasonable to leave this one out?

Reviewer: I probably miss-understood, then. I thought that today in SU meeting you said that we were not doing the multiple Artifacts because it got too complicated
